// contentful-migration/built/lib/intent/editor-layout/editor-layout-update-field-group.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_intent_1 = __importDefault(require("../base-intent"));
const chalk_1 = __importDefault(require("chalk"));
const editor_layout_update_field_group_1 = require("../../action/editor-layout/editor-layout-update-field-group");
const lodash_1 = require("lodash");
class EditorLayoutUpdateFieldGroupIntent extends base_intent_1.default {
    isEditorInterfaceIntent() {
        return true;
    }
    isEditorLayoutUpdate() {
        return true;
    }
    isFieldGroupUpdate() {
        return true;
    }
    isGroupable() {
        return true;
    }
    groupsWith(other) {
        return other.isGroupable()
            && other.isEditorLayoutUpdate()
            && this.isSameContentType(other);
    }
    endsGroup() {
        return false;
    }
    shouldSave() {
        return true;
    }
    shouldPublish() {
        return false;
    }
    toActions() {
        return [
            new editor_layout_update_field_group_1.EditorLayoutUpdateFieldGroupAction(this.payload.contentTypeId, this.payload.fieldGroupId, this.payload.fieldGroupProps)
        ];
    }
    toPlanMessage() {
        const details = (0, lodash_1.entries)(this.payload.fieldGroupProps).map(([key, value]) => {
            return (0, chalk_1.default) `{italic ${key}}: ${JSON.stringify(value)}`;
        });
        return {
            heading: (0, chalk_1.default) `Update editor layout for content type {bold.yellow ${this.getContentTypeId()}}`,
            sections: [{
                    heading: (0, chalk_1.default) `Update field group {yellow ${this.payload.fieldGroupId}}`,
                    details
                }],
            details: []
        };
    }
}
exports.default = EditorLayoutUpdateFieldGroupIntent;
