// contentful-migration/built/lib/intent/editor-layout/editor-layout-delete.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_intent_1 = __importDefault(require("../base-intent"));
const chalk_1 = __importDefault(require("chalk"));
const editor_layout_delete_1 = require("../../action/editor-layout/editor-layout-delete");
class EditorLayoutDeleteIntent extends base_intent_1.default {
    isEditorInterfaceIntent() {
        return true;
    }
    isEditorLayoutDelete() {
        return true;
    }
    isGroupable() {
        return false;
    }
    groupsWith() {
        return false;
    }
    endsGroup() {
        return true;
    }
    shouldSave() {
        return true;
    }
    shouldPublish() {
        return false;
    }
    toActions() {
        return [
            new editor_layout_delete_1.EditorLayoutDeleteAction(this.payload.contentTypeId)
        ];
    }
    toPlanMessage() {
        return {
            heading: (0, chalk_1.default) `Delete editor layout for content type {bold.yellow ${this.getContentTypeId()}}`,
            sections: [],
            details: []
        };
    }
}
exports.default = EditorLayoutDeleteIntent;
