// contentful-cli/dist/lib/core/event-handlers/intents/create-space-handler.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const { CREATE_SPACE_HANDLER } = require('../../events/scopes');
const inquirer = require('inquirer');
const createCreateSpaceHandlerIntents = () => {
    const createSpaceHandlerIntents = {
        scopes: [CREATE_SPACE_HANDLER],
        intents: {
            SELECT_ORG: ({ organizations }) => __awaiter(void 0, void 0, void 0, function* () {
                const answersOrganizationSelection = yield inquirer.prompt([
                    {
                        type: 'list',
                        name: 'organizationId',
                        message: 'Please select an organization:',
                        choices: organizations
                    }
                ]);
                return answersOrganizationSelection.organizationId;
            })
        }
    };
    return createSpaceHandlerIntents;
};
module.exports = createCreateSpaceHandlerIntents;
