// contentful-cli/dist/lib/cmds/content-type_cmds/list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const Table = require('cli-table3');
const { createManagementClient } = require('../../utils/contentful-clients');
const { handleAsyncError: handle } = require('../../utils/async');
const { log } = require('../../utils/log');
const paginate = require('../../utils/pagination');
const { highlightStyle } = require('../../utils/styles');
const { getHeadersFromOption } = require('../../utils/headers');
const { copyright } = require('../../utils/copyright');
module.exports.command = 'list';
module.exports.desc = 'List your content types';
module.exports.aliases = ['ls'];
module.exports.builder = yargs => {
    return yargs
        .option('space-id', { alias: 's', type: 'string', describe: 'Space id' })
        .option('management-token', {
        alias: 'mt',
        type: 'string',
        describe: 'Contentful management API token'
    })
        .option('environment-id', {
        alias: 'e',
        type: 'string',
        describe: 'Environment ID you want to interact with. Defaults to the current active environment.'
    })
        .option('header', {
        alias: 'H',
        type: 'string',
        describe: 'Pass an additional HTTP Header'
    })
        .epilog(copyright);
};
function ctList({ header, context }) {
    return __awaiter(this, void 0, void 0, function* () {
        const { managementToken, activeSpaceId, activeEnvironmentId } = context;
        const client = yield createManagementClient({
            accessToken: managementToken,
            feature: 'content_type-list',
            headers: getHeadersFromOption(header)
        });
        const space = yield client.getSpace(activeSpaceId);
        const environment = yield space.getEnvironment(activeEnvironmentId);
        log(highlightStyle(`Environment: "${activeEnvironmentId}"`));
        const result = yield paginate({
            client: environment,
            method: 'getContentTypes',
            query: {
                order: 'name,sys.id'
            }
        });
        const table = new Table({
            head: ['Content Type Name', 'Content Type ID']
        });
        result.items.forEach(contentType => {
            table.push([contentType.name, contentType.sys.id]);
        });
        log(table.toString());
    });
}
module.exports.handler = handle(ctList);
