// contentful-batch-libs/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.agentFromProxy = exports.proxyStringToObject = exports.writeErrorLogFile = exports.setupLogging = exports.displayErrorLog = exports.logEmitter = exports.wrapTask = exports.getEntityName = exports.addSequenceHeader = void 0;
var add_sequence_header_1 = require("./add-sequence-header");
Object.defineProperty(exports, "addSequenceHeader", { enumerable: true, get: function () { return __importDefault(add_sequence_header_1).default; } });
var get_entity_name_1 = require("./get-entity-name");
Object.defineProperty(exports, "getEntityName", { enumerable: true, get: function () { return __importDefault(get_entity_name_1).default; } });
var listr_1 = require("./listr");
Object.defineProperty(exports, "wrapTask", { enumerable: true, get: function () { return listr_1.wrapTask; } });
var logging_1 = require("./logging");
Object.defineProperty(exports, "logEmitter", { enumerable: true, get: function () { return logging_1.logEmitter; } });
Object.defineProperty(exports, "displayErrorLog", { enumerable: true, get: function () { return logging_1.displayErrorLog; } });
Object.defineProperty(exports, "setupLogging", { enumerable: true, get: function () { return logging_1.setupLogging; } });
Object.defineProperty(exports, "writeErrorLogFile", { enumerable: true, get: function () { return logging_1.writeErrorLogFile; } });
var proxy_1 = require("./proxy");
Object.defineProperty(exports, "proxyStringToObject", { enumerable: true, get: function () { return proxy_1.proxyStringToObject; } });
Object.defineProperty(exports, "agentFromProxy", { enumerable: true, get: function () { return proxy_1.agentFromProxy; } });
