// contentful-cli/dist/lib/cmds/space_cmds/delete.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const { confirmation } = require('../../utils/actions');
const { handleAsyncError: handle } = require('../../utils/async');
const { createManagementClient } = require('../../utils/contentful-clients');
const { copyright } = require('../../utils/copyright');
const { successEmoji } = require('../../utils/emojis');
const { getHeadersFromOption } = require('../../utils/headers');
const { log, success } = require('../../utils/log');
const { highlightStyle, warningStyle } = require('../../utils/styles');
module.exports.command = 'delete';
module.exports.desc = 'Deletes a space';
module.exports.builder = yargs => {
    return yargs
        .usage('Usage: contentful space delete')
        .option('space-id', {
        alias: 's',
        describe: 'ID of the space to delete',
        nargs: 1,
        demandOption: true
    })
        .option('management-token', {
        alias: 'mt',
        describe: 'Contentful management API token',
        type: 'string'
    })
        .option('yes', {
        describe: 'Skip the confirmation question'
    })
        .option('header', {
        alias: 'H',
        type: 'string',
        describe: 'Pass an additional HTTP Header'
    })
        .epilog(copyright);
};
function spaceDelete({ context, yes, header }) {
    return __awaiter(this, void 0, void 0, function* () {
        const { managementToken, activeSpaceId } = context;
        const client = yield createManagementClient({
            accessToken: managementToken,
            headers: getHeadersFromOption(header)
        });
        const space = yield client.getSpace(activeSpaceId);
        if (!yes) {
            log(`You are about to delete your ${highlightStyle(space.name)} (${highlightStyle(space.sys.id)}) space and all of its content. This cannot be undone.`);
            log();
            const confirm = yield confirmation('Do you want to confirm the deletion?');
            log();
            if (!confirm) {
                log(warningStyle(`Space deletion aborted.`));
                return;
            }
        }
        log();
        yield space.delete();
        log();
        success(`${successEmoji} Your ${highlightStyle(space.name)} (${highlightStyle(space.sys.id)}) space was successfully deleted.`);
    });
}
module.exports.spaceDelete = spaceDelete;
module.exports.handler = handle(spaceDelete);
