// contentful-cli/dist/lib/cmds/init/success.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const cli_highlight_1 = require("cli-highlight");
const FEEDBACK_LINK = 'https://contentful.typeform.com/contentful-cli';
function success({ accessToken, connectionType, environmentId = '{environment_id}', spaceId = '{space_id}' }) {
    return __awaiter(this, void 0, void 0, function* () {
        const examples = {
            'Contentful.js': (0, cli_highlight_1.highlight)(`
    import contentful from "contentful";

    const client = contentful.createClient({
      space: "${spaceId}",
      accessToken: "${accessToken}"
    });

    client.getEntries()
      .then((response) => console.log(response.items))
      .catch(console.error)
    `, { language: 'JavaScript' }),
            GraphQL: `
    curl -g \\
      -X POST \\
      -H "Content-Type: application/json" \\
      -H "Authorization: Bearer ${accessToken}" \\
      -d  '{"query":"query($preview:Boolean){entryCollection(preview:$preview){items{__typename}}}","variables":{"preview":true}}' \\
      https://graphql.contentful.com/content/v1/spaces/${spaceId}/environments/${environmentId}
      `,
            'REST API': `
    curl --include \\
      --request GET \\
      https://preview.contentful.com/spaces/${spaceId}/environments/${environmentId}/entries?access_token=${accessToken}
    `,
            React: (0, cli_highlight_1.highlight)(`
    export const useContentfulQuery = ({ query, variables }) => {
      const [data, setData] = useState(null);
      useEffect(() => {
        if (!query || !variables) return;
        const getContentfulData = async () => {
          const response = await fetch("https://graphql.contentful.com/content/v1/spaces/${spaceId}",
            {
              method: "POST",
              headers: {
                "Content-Type": "application/json",
                Authorization: "Bearer ${accessToken}",
              },
              body: JSON.stringify({ query, variables,}),
            }
          );
          const { data: contentfulData } = await response.json();
          setData(contentfulData);
        };
        getContentfulData();
      }, [query, variables]);
      return { data };
    };

    // Then in your component:
    const { data } = useContentfulQuery({
      query: "query($preview:Boolean){entryCollection(preview:$preview){items{__typename}}}",
      variables: { preview: true },
    });
    `, { language: 'JavaScript' })
        };
        console.log(chalk_1.default.cyan(`\n-----Try your first request-------------------------------------------------------\n`), examples[connectionType], chalk_1.default.cyan(`\n----------------------------------------------------------------------------------\n`));
        switch (connectionType) {
            case 'React':
            case 'GraphQL':
                console.log(`Use ${chalk_1.default.magenta('GraphiQL')} Playground app and start building your queries: ${chalk_1.default.blue('https://app.contentful.com/deeplink?link=apps-contentful&id=graphql-playground\n')}`, `\nLearn more about GraphQL: ${chalk_1.default.blue('https://contentful.com/developers/docs/references/graphql/')}\n`);
                break;
            case 'Contentful.js':
                console.log(`Learn more about ${chalk_1.default.yellow('Content Delivery API JS')} client library: ${chalk_1.default.blue('https://www.contentful.com/developers/docs/javascript/sdks')}\n`);
                break;
            case 'REST API':
                console.log(`Learn more about Content Delivery API: ${chalk_1.default.blue('https://www.contentful.com/developers/docs/references/content-delivery-api')}\n`);
                break;
        }
        console.log(`Share your feedback for this CLI: ${chalk_1.default.blue(FEEDBACK_LINK)}\n`);
    });
}
exports.default = success;
