// contentful-cli/dist/lib/utils/merge/render-changeset-messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderChangesetMessages = void 0;
const spacer = ' ';
function StringRenderer({ indentString = spacer.repeat(2) }) {
    const lines = [];
    return {
        addLine: (text, indent) => {
            return lines.push(`${indentString.repeat(indent)}${text}`);
        },
        toString: () => lines.join('\n')
    };
}
const ChangeTypeToIndicatorMap = {
    add: '+Added',
    delete: '-Deleted',
    update: '~Changed',
    move: '~Changed'
};
function renderChangesetMessages(messages) {
    const renderer = StringRenderer({});
    for (const message of messages) {
        renderer.addLine(`${ChangeTypeToIndicatorMap[message.changeType]}`, message.level);
        for (const messageLine of message.messages) {
            renderer.addLine(`${spacer}${messageLine}`, message.level);
        }
        // add empty line between every block
        renderer.addLine('', 0);
    }
    return renderer.toString();
}
exports.renderChangesetMessages = renderChangesetMessages;
