// contentful-cli/dist/lib/cmds/extension_cmds/get.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const { createManagementClient } = require('../../utils/contentful-clients');
const { handleAsyncError: handle } = require('../../utils/async');
const { logExtension } = require('./utils/log-as-table');
const { getHeadersFromOption } = require('../../utils/headers');
const { copyright } = require('../../utils/copyright');
module.exports.command = 'get';
module.exports.desc = 'Show an extension';
module.exports.builder = yargs => {
    return yargs
        .option('id', { type: 'string', demand: true, describe: 'Extension id' })
        .option('management-token', {
        alias: 'mt',
        type: 'string',
        describe: 'Contentful management API token'
    })
        .option('space-id', { type: 'string', describe: 'Space id' })
        .option('environment-id', { type: 'string', describe: 'Environment id' })
        .option('header', {
        alias: 'H',
        type: 'string',
        describe: 'Pass an additional HTTP Header'
    })
        .epilog(copyright);
};
function getExtension({ context, id, header }) {
    return __awaiter(this, void 0, void 0, function* () {
        const { managementToken, activeSpaceId, activeEnvironmentId } = context;
        const client = yield createManagementClient({
            accessToken: managementToken,
            feature: 'extension-get',
            headers: getHeadersFromOption(header)
        });
        const space = yield client.getSpace(activeSpaceId);
        const environment = yield space.getEnvironment(activeEnvironmentId);
        const extension = yield environment.getUiExtension(id);
        logExtension(extension, activeSpaceId, activeEnvironmentId);
    });
}
module.exports.handler = handle(getExtension);
