// contentful-cli/dist/lib/cmds/extension_cmds/list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const Table = require('cli-table3');
const { createManagementClient } = require('../../utils/contentful-clients');
const { handleAsyncError: handle } = require('../../utils/async');
const { log } = require('../../utils/log');
const paginate = require('../../utils/pagination');
const { getHeadersFromOption } = require('../../utils/headers');
const { copyright } = require('../../utils/copyright');
module.exports.command = 'list';
module.exports.desc = 'List all extensions';
module.exports.builder = yargs => {
    return yargs
        .option('management-token', {
        alias: 'mt',
        type: 'string',
        describe: 'Contentful management API token'
    })
        .option('space-id', { type: 'string', describe: 'Space id' })
        .option('environment-id', { type: 'string', describe: 'Environment id' })
        .option('header', {
        alias: 'H',
        type: 'string',
        describe: 'Pass an additional HTTP Header'
    })
        .epilog(copyright);
};
function listExtensions({ context, header }) {
    return __awaiter(this, void 0, void 0, function* () {
        const { managementToken, activeSpaceId, activeEnvironmentId } = context;
        const client = yield createManagementClient({
            accessToken: managementToken,
            feature: 'extension-list',
            headers: getHeadersFromOption(header)
        });
        const space = yield client.getSpace(activeSpaceId);
        const environment = yield space.getEnvironment(activeEnvironmentId);
        const result = yield paginate({
            client: environment,
            method: 'getUiExtensions',
            query: {
                order: 'extension.name,sys.id'
            }
        });
        if (result.items.length) {
            const extensions = result.items.sort((a, b) => a.extension.name.localeCompare(b.extension.name));
            const table = new Table({
                head: ['Extension Name', 'Extension ID', 'Version']
            });
            extensions.forEach(extension => {
                table.push([
                    extension.extension.name,
                    extension.sys.id,
                    extension.sys.version
                ]);
            });
            log(table.toString());
            return;
        }
        log('No extensions found');
    });
}
module.exports.listExtensions = listExtensions;
module.exports.handler = handle(listExtensions);
