// contentful-cli/dist/lib/cmds/init.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.init = exports.builder = exports.desc = exports.command = void 0;
const inquirer_1 = __importDefault(require("inquirer"));
const async_1 = require("../utils/async");
const greetings_1 = __importDefault(require("./init/greetings"));
const success_1 = __importDefault(require("./init/success"));
const context_1 = require("../context");
const login_1 = require("./login");
const chalk_1 = __importDefault(require("chalk"));
const apikey_1 = require("./init/apikey");
const space_1 = require("./init/space");
exports.command = 'init';
exports.desc = 'Get started with Contentful';
const builder = (yargs) => {
    return yargs
        .usage('Start a new project or see how to interact with an existing one!')
        .epilog('Your Contentful Guide: https://contentful.com/developers');
};
exports.builder = builder;
const init = () => __awaiter(void 0, void 0, void 0, function* () {
    (0, greetings_1.default)();
    const context = yield (0, context_1.getContext)();
    if (!context.managementToken) {
        context.managementToken = yield (0, login_1.login)({ context });
    }
    if (!context.managementToken)
        return console.log(chalk_1.default.red("We couldn't find your access token, please login again!"));
    const space = yield (0, space_1.getSpace)(context);
    const environmentId = (yield space.getEnvironments()).items[0].sys.id || 'master';
    const apiKey = yield (0, apikey_1.getPreviewApiKey)(space, environmentId);
    const { connectionType } = yield inquirer_1.default.prompt({
        type: 'list',
        name: 'connectionType',
        prefix: '👓',
        message: 'How would you like to consume content in your code?',
        choices: [
            {
                name: `${chalk_1.default.blue('React')}`,
                value: 'React'
            },
            {
                name: `${chalk_1.default.magenta('GraphQL')}`,
                value: 'GraphQL'
            },
            {
                name: `${chalk_1.default.yellow('JS Client')}`,
                value: 'Contentful.js'
            },
            {
                name: 'REST API',
                value: 'REST API'
            }
        ]
    });
    (0, success_1.default)({
        accessToken: apiKey,
        connectionType,
        spaceId: space.sys.id,
        environmentId
    });
});
exports.init = init;
exports.handler = (0, async_1.handleAsyncError)(exports.init);
