// contentful-cli/dist/lib/cmds/login.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.login = exports.builder = exports.desc = exports.command = exports.getOauthURL = void 0;
const open_1 = __importDefault(require("open"));
const inquirer_1 = __importDefault(require("inquirer"));
const chalk_1 = __importDefault(require("chalk"));
const context_1 = require("../context");
const actions_1 = require("../utils/actions");
const async_1 = require("../utils/async");
const styles_1 = require("../utils/styles");
const token_info_1 = require("../utils/token-info");
const copyright_1 = require("../utils/copyright");
const APP_ID = '9f86a1d54f3d6f85c159468f5919d6e5d27716b3ed68fd01bd534e3dea2df864';
const getOauthURL = (host) => {
    const REDIRECT_URI = 'https://www.contentful.com/developers/cli-oauth-page/';
    let oAuthURL = `https://be.contentful.com/oauth/authorize?response_type=token&client_id=${APP_ID}&redirect_uri=${REDIRECT_URI}&scope=content_management_manage&action=cli`;
    if (host) {
        const url = host.replace('api.', '');
        oAuthURL = `https://be.${url}/oauth/authorize?response_type=token&client_id=${APP_ID}&redirect_uri=${REDIRECT_URI}&scope=content_management_manage&action=cli`;
    }
    return oAuthURL;
};
exports.getOauthURL = getOauthURL;
exports.command = 'login';
exports.desc = 'Login to Contentful';
const builder = (yargs) => yargs
    .usage('Usage: contentful login')
    .option('management-token', {
    alias: 'mt',
    describe: 'Contentful management API token',
    type: 'string'
})
    .epilog([
    'See more at:',
    'https://github.com/contentful/contentful-cli/tree/master/docs/login',
    copyright_1.copyright
].join('\n'));
exports.builder = builder;
const login = ({ context, managementToken: managementTokenFlag }) => __awaiter(void 0, void 0, void 0, function* () {
    const { managementToken, host } = context;
    let token;
    if (managementTokenFlag) {
        token = managementTokenFlag;
    }
    else {
        if (managementToken) {
            console.log(`You're already logged in!`);
            yield (0, token_info_1.tokenInfo)();
            console.log(`To logout, type: ${chalk_1.default.green('contentful')} ${chalk_1.default.magenta('logout')}\n`);
            return managementToken;
        }
        console.log(`A browser window will open where you will log in (or sign up if you don’t have an account), authorize this CLI tool and paste your ${(0, styles_1.highlightStyle)('CMA token')} here:\n`);
        const confirmed = yield (0, actions_1.confirmation)('Continue login on the browser?');
        if (!confirmed) {
            console.log(chalk_1.default.red('Aborted!'), 'please login to use Contentful CLI features!', `\nUsage: ${chalk_1.default.green('contentful')} ${chalk_1.default.cyan('login')}`);
            return;
        }
        const oAuthURL = (0, exports.getOauthURL)(host);
        // We open the browser window only on Windows and OSX since this might fail or open the wrong browser on Linux.
        if (['win32', 'darwin'].includes(process.platform)) {
            yield (0, open_1.default)(oAuthURL, {
                wait: false
            });
        }
        else {
            console.log(`Unable to open your browser automatically. Please open the following URI in your browser:\n\n${(0, styles_1.pathStyle)(oAuthURL)}\n\n`);
        }
        const tokenAnswer = yield inquirer_1.default.prompt([
            {
                type: 'password',
                mask: true,
                name: 'managementToken',
                message: 'Paste your token here:',
                validate: val => /^[a-zA-Z0-9_-]{43,64}$/i.test(val.trim()) // token is 43 to 64 characters and accepts lower/uppercase characters plus `-` and `_`
            }
        ]);
        token = tokenAnswer.managementToken;
    }
    yield (0, context_1.setContext)({
        managementToken: token
    });
    yield (0, context_1.storeRuntimeConfig)();
    console.log(`\n${chalk_1.default.green('Great!')} You've successfully logged in!`);
    yield (0, token_info_1.tokenInfo)();
    return token;
});
exports.login = login;
exports.handler = (0, async_1.handleAsyncError)(exports.login);
